Trashbin endpoint in DAV

rullzer opened the pull request:
Fixes #1332


- [x] Add a new DAV collection
- [x] List all files in the trashbin for this user
- [x] Deleting files from trashbin
- [x] Get files from trashbin (just read)
- [x] Empty trashbin (just a delete on it)
- [x] Restore from trashbin 
   * To special restore node that will maitain current behavior (with conflict handling etc).
- [x] Add dav properties
  - [x] Original name
  - [x] Original location

Future work:
-  Restore from trashbin
  * Manual move to files
-  Migrate web UI to DAV
-  Add integration tests

rullzer committed:
List trashbin in DAV

First steps for #1332

* Add a new DAV collection
* List all files in the trashbin for this user
* Deleting files from trashbin
* Get files from trashbin (just read)

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Emptry trashbin

A delete on a users trashbin will empty it (the trashbin won't be
deleted of course)

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add special restore folder

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer committed:
Add propfind properties to trashbin

* get original filename (without the weird timestamp)
* get original location

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
Ready for review I would say.

Basic functionality is there.

MorrisJobke commented on apps/dav/lib/RootCollection.php:
```diff
@@ -55,7 +55,7 @@ public function __construct() {
 		$groupPrincipalBackend = new GroupPrincipalBackend($groupManager);
 		// as soon as debug mode is enabled we allow listing of principals
-		$disableListing = !$config->getSystemValue('debug', false);
+		$disableListing = false; //!$config->getSystemValue('debug', false);
```
Is this wanted?

rullzer replied:
Ah no... I'll fix that :S

rullzer replied:
fixed

icewind1991 commented on apps/files_trashbin/lib/Sabre/TrashFolder.php:
```diff
@@ -0,0 +1,116 @@
+
+		foreach ($entries as $entry) {
+			if ($entry->getName() === $name) {
+				if ($entry->getMimetype() === 'httpd/unix-directory') {
```
`$entry->isDirectory()`

rullzer replied:
Well that doesn't exist on https://github.com/nextcloud/server/blob/master/lib/public/Files/FileInfo.php ;)

rullzer replied:
But I can use getType

icewind1991 replied:
Confused it with `IFileInfo` from the smb lib again :see_no_evil:

icewind1991 commented on apps/files_trashbin/lib/Sabre/TrashFolder.php:
```diff
@@ -0,0 +1,116 @@
+	}
+
+	public function childExists($name) {
+		$entries = \OCA\Files_Trashbin\Helper::getTrashFiles($this->getName(), $this->userId);
```
this seems horribly inefficient

rullzer replied:
It is. But good enough for now

rullzer replied:
Basically this code is now doing pretty much the same as the original. Baby steps we can optimize later.

MorrisJobke reviewed (approved):
Tested and works :+1:

icewind1991 reviewed (dismissed):
Also I would prefer to have the trash listing in `trashbin/$user/trash` so the restore folder doesn't have to be in the trash listing

MorrisJobke commented:
CI failure unrelated

rullzer committed:
Use getType to determine type of entry

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
@icewind1991 `trashbin/$user/trash` looks weird to me... But I get the point... @MorrisJobke opinions?

MorrisJobke commented:
> @icewind1991 trashbin/$user/trash looks weird to me... But I get the point... @MorrisJobke opinions?

Makes sense 👍

rullzer commented:
OK let me fix that later this evening

rullzer committed:
Add folder for trashbin

The trashbin home now contains 2 entries restore and trash.

Made all files strict. Added more types.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

rullzer commented:
@icewind1991 please have another look

rullzer commented:
@icewind1991 ping
